Add a dark theme

prymitive opened the pull request:
This implements dark mode theme using https://github.com/ForEvolve/bootstrap-dark.
User is able to set theme to:
a) light - same as current theme and the default one
b) dark - dark version using bootstrap-dark
c) auto - either light or dark version will be used depending on browser preference, if browser doesn't support media queries for prefers-color-scheme light theme will be used

To make all of this work existing CSS rules were moved into themes folder and converted to scss, both light and dark styles will import that shared CSS file but each might pass different variables to customise generated CSS. Some parts of the shared CSS file use bootstrap and bootstrap-dark variables directly, to avoid redefining what's already defined there.

Signed-off-by: Łukasz Mierzwa <l.mierzwa@gmail.com>

roidelapluie commented:
Thanks for this!!

I can't build it it seems:

```
cd web/ui/react-app && yarn --frozen-lockfile
yarn install v1.22.5
[1/4] Resolving packages...
[2/4] Fetching packages...
info fsevents@2.1.2: The platform "linux" is incompatible with this module.
info "fsevents@2.1.2" is an optional dependency and failed compatibility check. Excluding it from installation.
info fsevents@1.2.13: The platform "linux" is incompatible with this module.
info "fsevents@1.2.13" is an optional dependency and failed compatibility check. Excluding it from installation.
info fsevents@2.3.2: The platform "linux" is incompatible with this module.
info "fsevents@2.3.2" is an optional dependency and failed compatibility check. Excluding it from installation.
[3/4] Linking dependencies...
warning " > eslint-config-react-app@5.2.1" has unmet peer dependency "babel-eslint@10.x".
[4/4] Building fresh packages...
Done in 41.77s.
>> building React app
building React app
yarn run v1.22.5
$ react-scripts build
Creating an optimized production build...
Failed to compile.

./src/App.tsx
Cannot find file './contexts/ThemeContext' in './src'.


error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

prymitive commented:
Missed one file when making the PR, should be there now

roidelapluie commented:
Thanks! This looks good, some elements are still bright and not adapted, but from an end user perspective, I like this.

prymitive commented:
I’d hope I’ll be able to adopt more elements to bootstrap styles, like query autocomplete, which would improve consistency and reduce amount of custom css needed.

juliusv commented:
@prymitive Awesome, will take a look!

Don't put any energy into the expression input field yet, I'm going to replace that in the next weeks with the new and awesome PromQL editor functionality from https://github.com/prometheus-community/codemirror-promql and PromLens.

juliusv commented:
One thing I noticed: switching themes while having a graph open makes the app crash like this:

![crash](https://user-images.githubusercontent.com/538008/111228981-6ac17380-85e5-11eb-9664-5503572c1031.png)

Seems like the graph container height briefly becomes 0 during the theme switch, and then `handleResize()` chokes on that.

juliusv commented:
Also got this error sometimes when trying to graph an expression, not sure if it's related:

![crash2](https://user-images.githubusercontent.com/538008/111229215-cee43780-85e5-11eb-9e6b-d19d2d5d4287.png)

prymitive commented:
Yes, I've noticed some issues with that as well, might need to make some css theme-free or look closer into how height is being calculated.

prymitive commented:
Moved most of the css back, except for those rules that are templated via scss variables. Hopefully this will prevent element sizes from resetting on theme swap.

juliusv commented:
Cool! Just some more cursory testing for now:

- Still got the `Cannot read property 'save' of null` error above once, when graphing `prometheus_http_request_duration_seconds_count`, but I'm not yet sure how to reproduce it again.
- The graph time control input fields (range, etc.) are now all the same width, whereas they had custom widths before.

prymitive commented:
Yeah, I do see failures too, will put this on hold as a draft and will add storybook first so once this is ready we can use storybook to compare all UI elements and see differences between current UI and new themes.

prymitive commented:
> Still got the Cannot read property 'save' of null error above once, when graphing prometheus_http_request_duration_seconds_count, but I'm not yet sure how to reproduce it again.

I think what happens here is that when you toggle themes:
- some styles reset
- that trigger graph size change
- which triggers resize detector
- which triggers `handleResize()`
- which calls plot()
- which triggers a graph redraw via setTimeout() 
- if there's another re-render before setTimeout() triggered flot happens it might destroy the canvas and once graph redraw finally runs it gets a null

Or so it seems so. The short term fix is to fix styles so there is no graph resize on theme change.

prymitive commented:
I think that most of the issues are due to lazy loading theme css.
The effect is that upon first render there is no bootstrap style applied at all, since body now must have either bootstrap or bootstrap-dark class to use bootstrap. Those classes are applied in Theme component that also lazy loads needed css file, so on first render we have body with no class and no styles, we render a bunch of components, in the mean time css loads and all styles are applied, which triggers a bunch of resizes and other updated, and that isn't all handled well.
This is mostly visible on the graph page where:
* input gets wrong size - there some js code that set fixed size via style attributes, with lazy loaded css we end up with height of an unstyled input, which is less that what bootstrap will have
* flot charts - we see crashes, which seems it's related to some re-renders that are done via setTimeout in the flot lib

So for now I've disabled lazy loading, both themes are loaded with App.css and also added bootstrap class to body, so it uses light theme by default. This should mitigate those issues for now and we'll be able work on making dark theme usable.
We can address lazy loading and graph page issues caused by it later.

prymitive commented:
Cleaned up the PR a bit and hopefully the UI looks good now.
Lazy loading for theme css would mean 25KB of css instead of 50, but it also introduces an extra step when it all loads and there's a spinner for split seconds, so getting rid of it will make the UX a bit better, so it's not all bad that we can't have it.
Need to add some tests and finish storybook work so we can see all UI elements in various states for both themes easily.

juliusv commented on web/ui/react-app/src/Theme.tsx:
```diff
@@ -0,0 +1,48 @@
+        </Button>
+        <Button
+          color="secondary"
+          title="Use browser preferred theme"
```
Nit: Either

```suggestion
          title="Use browser's preferred theme"
```

or (preferred, compound adjective form, http://www.kentlaw.edu/academics/lrw/grinker/LwtaCompound_Adjectives.htm):

```suggestion
          title="Use browser-preferred theme"
```

juliusv commented on web/ui/react-app/src/pages/graph/ExpressionInput.tsx:
```diff
@@ -215,12 +215,10 @@ class ExpressionInput extends Component<ExpressionInputProps, ExpressionInputSta
                 />
                 <InputGroupAddon addonType="append">
-                  <Button className="btn-light border" title="Open metrics explorer" onClick={this.openMetricsExplorer}>
+                  <Button title="Open metrics explorer" onClick={this.openMetricsExplorer}>
```
Due to the removal of `btn-light`, this button is now always dark, even in light mode.

prymitive replied:
This one is a bit tricky since there are no border or other visual separators for buttons in a button group

https://getbootstrap.com/docs/4.6/components/button-group/

Which means that when you have a button group with buttons of different colour it tends to look a bit "raw", at least to my eye. I'll try to add some custom style to it.

juliusv commented on web/ui/react-app/src/themes/light.scss:
```diff
@@ -0,0 +1,15 @@
+@import '~bootstrap/scss/variables';
+
+$config-yaml-color: #333;
+$config-yaml-bg: $jumbotron-bg;
```
Rather than repurposing `$jumbotron-bg` (whose name is confusing in this context, and whose color is a bit darker than what we used here previously), I think let's use a custom color?

juliusv commented on web/ui/react-app/src/themes/shared.scss:
```diff
@@ -0,0 +1,40 @@
+  border-bottom: 1px solid $nav-tabs-border-color;
+}
+
+.rule_cell {
```
Let's use the opportunity to make the old under_scored CSS class names (`rule_cell` and `alert_cell`) consistent with the dash-separated ones :)

juliusv reviewed (commented):
Looking pretty great already! :) Just a couple of nits, and of course at least some minor tests would be good to have before merging.

I still got the `Cannot read property 'save' of null error` once by the way, so I'm not sure whether it's really related to the lazy loading (now that you've removed that).

juliusv commented:
Note: This will need a rebase, we just merged the new expression editor yesterday: https://github.com/prometheus/prometheus/pull/8634

I can do the dark themeing for the editor itself in the end (otherwise you have to learn CodeMirror styling specifics), so don't worry about that part.

prymitive commented:
Rebased and addressed (most?) comments

juliusv commented on web/ui/react-app/src/pages/graph/ExpressionInput.tsx:
```diff
@@ -215,12 +215,10 @@ class ExpressionInput extends Component<ExpressionInputProps, ExpressionInputSta
                 />
                 <InputGroupAddon addonType="append">
-                  <Button className="btn-light border" title="Open metrics explorer" onClick={this.openMetricsExplorer}>
+                  <Button className="metrics-explorer" title="Open metrics explorer" onClick={this.openMetricsExplorer}>
```
Nice, I like how it looks now! The class name is a bit confusing since `metrics-explorer` is also used for the metrics explorer itself. Maybe `metrics-explorer-btn` or something like that would be clearer?

prymitive replied:
I was hoping that being able to target it with `button.foo` is enough to make it clear but let's rename it for clarity

juliusv replied:
Yep, I just think the class name itself would have been a bit of a misnomer otherwise... thanks!

juliusv commented:
@prymitive So now we have to collaborate somehow on bringing the new editor styling into this. I can't push to your private PR branch, but one option would be to open a short-lived feature branch in Prometheus itself where we can both merge stuff into after each other? I'd just like to avoid merging it direclty into `main` with known brokenness for the new editor.

Or alternatively, if you feel bored (because today I won't have much time at least): Basically what needs to be done is to apply some of the same input group styling changes to the new editor as well, and then update the CodeMirror theme + highlighting colors in https://github.com/prometheus/prometheus/blob/main/web/ui/react-app/src/pages/graph/CMTheme.tsx to add a dark theme. Here's some examples from the base theme in CodeMirror for how to target light and dark theme selectors: https://github.com/codemirror/view/blob/main/src/theme.ts#L71-L72. And then the `EditorView.theme()` call can take a second parameter of the form `{dark: <boolean>}` to indicate to CodeMirror whether to style the editor with the dark theme or not.

juliusv commented:
Btw., the `.expression-input .cm-expression-input` styling from https://github.com/prometheus/prometheus/pull/8634/files seems to have not made it over into here in the rebase.

prymitive commented:
You should be able to push to my branch:

<img width="271" alt="Screenshot 2021-03-24 at 11 15 33" src="https://user-images.githubusercontent.com/114288/112301934-61de3b00-8c92-11eb-991c-8951a927ff65.png">

Let's resume tomorrow, I don't have much more time left for this today anyway.

juliusv commented:
Ah ok, cool! Then I will do that, with your permission :) And yep. Will see how much time I find today, otherwise tomorrow.

prymitive commented:
> Btw., the `.expression-input .cm-expression-input` styling from https://github.com/prometheus/prometheus/pull/8634/files seems to have not made it over into here in the rebase.

I must have missed it while rebasing or picked wrong change when doing so, should be there now

prymitive commented:
FYI I can look into theming new input over weekend

prymitive commented:
Since theming for CMExpressionInput is done in JS, where I can't easily use SCSS variables, I migrated it to scss.
It should work fairly consistent with the rest of web ui.
There are still styles generated and injected by codemirror that would be good to get rid of but I'm not sure if I can instruct it to "use empty base theme", it seems there's always default bundled and a way of overriding it. Or at least I'm not seeing a way of getting rid of default theme for it.

juliusv commented:
@prymitive Hmm, TBH I really liked how the editor styling was so encapsulated, but I admit I also don't have super good arguments for keeping it in TypeScript. Maybe just because that's the usual CodeMirror way, and it makes it easier for other people to extract & reuse just the editor. But I guess it's fine to move it to SCSS if that makes variable usage that much easier.

But it seems that the style migration changed the appearance even in the light theme, which I would like to keep the same:

PR:

![pr-theme](https://user-images.githubusercontent.com/538008/112766039-41a2d900-9010-11eb-83e5-e727c0f5e39c.png)

vs. main:

![main-theme](https://user-images.githubusercontent.com/538008/112766044-4cf60480-9010-11eb-8a71-67a0d0e1ed31.png)

The completion item icons are also missing now.

I think we should also define different highlighter styles for light & dark themes, as the current highlighter produces pretty low contrast for some of the tokens in the dark theme (esp. label name and label value):

![lowcontrast](https://user-images.githubusercontent.com/538008/112766242-22587b80-9011-11eb-998e-a9427e0d42d7.png)

juliusv commented:
Thinking about this further, I guess I'm fine with slight style changes even in the light theme, as those are taken from Bootstrap-theme-related colors etc., and thus hopefully yield a more coherent styling with the rest of the UI. But I don't like some of the specifics (mostly the turquoise fuzzy-match highlight color and the missing icons, and I think it'd be nice to give the dropdown at least some off-white coloring to make it stand out from the background a bit).

But yeah, for the highlighter I don't think we can get around having two styles directly in TypeScript.

prymitive commented:
I think it all depends on what we're trying to achieve here.
I was aiming at visual consistency, hence trying to re-use bootstrap styles as much as possible so, for example, we don't have a different font heigh, colour and  styles in various components. Especially that it increases maintenance cost over time - it's easy to add a new component with some extra hardcoded styles, make it visually blend into current overall look of the app, but then you come to a point where you need to upgrade bootstrap (that defines 99% of look of the app) and suddenly some components start to look a bit off brand.
You can see this in the diff already, although I tried to make it as little invasive as possible, there's a non trivial number of places that needs styles re-adjusted when adding a new theme, because they were written for the default one and so they assume that all other base css is from it.
This is partially why I added scss to the mix - instead of copying colour from bootstrap theme to write some extra css we can import variables that define it and use to generate that extra css. This allows us to create extra components, not covered by bootstrap component library, that look and feel like bootstrap. This is also why I moved the theme of new input into scss - I cannot re-use bootstrap scss variables inside JS code.
If visual consistency is not a concern we skip those adjustments

I would personally try to minimise custom css and use as much as possible of bootstrap. Especially that it allows customising. IMHO using a framework like boostrap greatly reduces the amount of maintenance and, if you steer clear from adding tons of custom manual styles on top of it, upgrades to newer bootstrap are fairly easy.

juliusv commented:
@prymitive Yeah you're right, we should mostly stick to the light/dark Bootstrap styles, so let's go with the SCSS. I'd still like to make occasional exceptions to styles though, like the ones stated in my previous comment. With the code editor, it's a quite specialized environment that often calls for styling that isn't 100% "Bootstrap-like" IMO, because Bootstrap wasn't created with styling that kind of component in mind.

juliusv commented:
FYI: I made a small styling change to the expression editor in https://github.com/prometheus/prometheus/pull/8665 that will be merged into the 2.26 release branch, and then after the release, into the main branch. This PR will eventually need to include that.

Talador12 commented:
Request: could we update the dark mode page load order to show a dark screen while the page is loading? If you have it set to dark mode and it takes a minute to load, it shows a blinding white screen until the page is loaded.

prymitive commented:
> FYI: I made a small styling change to the expression editor in #8665 that will be merged into the 2.26 release branch, and then after the release, into the main branch. This PR will eventually need to include that.

Will look into that this weekend

prymitive commented:
Started cleaning and splitting changes into smaller chunks to avoid `App.css` being renamed, which made it harder to follow style changes. Got a few things to do but should be able to update this in a few days.

juliusv commented:
@prymitive Great. Yeah, hopefully we will have fewer (or no) other style changes to rebase on in the coming days / weeks, and once we get it merged, the pain is over :)

prymitive commented:
@juliusv I've pretty much redo the whole PR this should be now easier to follow
I kept style changes to minimum to add a dark theme rather then modify light one and use it as a base for dark mode, changes to default css are one where needed to branch style into light and dark

Nexucis commented on web/ui/react-app/package.json:
```diff
@@ -31,18 +32,20 @@
     "tempusdominus-bootstrap-4": "^5.1.2",
     "tempusdominus-core": "^5.0.3",
-    "typescript": "^3.3.3"
+    "typescript": "^3.3.3",
```
it could be nice to move the `typescript` dependency to the `devDependencies` section. And upgrade the typescript to v4 if it's possible. (yeah I'm chasing the dependency that should really be into the dependency section :D)

prymitive replied:
How is that related to the dark theme? I don’t think I follow

Nexucis replied:
oh it's not relative, I just saw the diff. So I was more talking about taking the opportunity to do it.

prymitive replied:
I don't think it's a good idea to inflate already large PR with unrelated changes.
I would suggest to open a dedicated PR for it. Especially that it, at least to me, it doesn't seem at all obvious why ( typescript is required to build the app) so a longer explanation there would be useful.

juliusv replied:
Yep let's leave that for a separate PR.

juliusv commented:
@prymitive Looks pretty good to me from a code perspective now, thanks! A couple of questions / notes still:

* It'll need a rebase to make PR conflicts go away.
* If I understand correctly, in the future, if I have anything that stays the same across themes, I would put it into `App.css`, but if it's something that can be customized, I would use `_shared.scss`. Or would it at some point make sense to move all of `App.css` into `_shared.scss` as well, and you just avoided that here to keep the PR more focused?
* The various time inputs on the graph lost their custom widths again.
* Some colors still seem off to me in dark mode: either too low contrast (e.g. dark-blue-on-black partial match highlighting in the autocomplete dropdown in the new editor) or too high contrast / non-fitting color mix (e.g. the "No data queried yet" `alert-light` on /graph). But I think this is fine for a start and it's probably easier for me (or others) to play with it and send in additional changes after this PR has been merged.

prymitive commented:
> * If I understand correctly, in the future, if I have anything that stays the same across themes, I would put it into `App.css`, but if it's something that can be customized, I would use `_shared.scss`. Or would it at some point make sense to move all of `App.css` into `_shared.scss` as well, and you just avoided that here to keep the PR more focused?

Yes. I kept `App.css` as is for now as moving that code is a one way street (as seen in my previous attempt), once we do it it's really hard to keep track of what was added, modified etc.
I think it would make sense, once this is all ready to merge, to move it to `_shared.scss`, especially that we need to nest all rules under `.bootstrap` & `.bootstrap-dark` class selectors to make sure that rules are always applied (otherwise there might be more specific selectors in bootstrap / components).

> * The various time inputs on the graph lost their custom widths again.

Missed that, will fix.

> * Some colors still seem off to me in dark mode: either too low contrast (e.g. dark-blue-on-black partial match highlighting in the autocomplete dropdown in the new editor) or too high contrast / non-fitting color mix (e.g. the "No data queried yet" `alert-light` on /graph). But I think this is fine for a start and it's probably easier for me (or others) to play with it and send in additional changes after this PR has been merged.

Agreed, as you can see on the commit history I tried to attack it one problem at a time. I might need to finish storybook work so we get a better visual comparison.

juliusv commented:
Cool, sounds good to me. Let me know when I should take another look then :)

prymitive commented:
Btw: what I also found was that if you press `ctrl+f`, while focusing input in the new editor, you'll get a search UI that's currently unstyled. Do we want that styled for both themes?

juliusv commented:
@prymitive We probably just want to remove the search plugin for CodeMirror altogether :)

roidelapluie commented:
I think we should get this merged once rebased (and custom width fixed) so we can address the few remaining issues individually (like some colors).

juliusv commented:
Agreed!

prymitive commented:
Rebased and move `App.css` to `_shared.scss` (which should fix width issues).

prymitive commented:
I need to move `html {}` style out of bootstrap selectors

juliusv commented:
> I need to move html {} style out of bootstrap selectors

@prymitive Does that mean I shouldn't merge yet?

prymitive commented:
I've noticed that there's a rule for `html {}` and that wouldn't be correctly applied from `_shared.scss`.
Just added an extra commit that moves it

juliusv commented:
Nice. One last nit I found before merging: the icons in buttons (metrics explorer button or time control buttons) were a couple of pixels higher in the existing UI. It seems that this styling for the icons got lost, and now they are a tad too low:

```
.svg-inline--fa {
    display: inline-block;
    font-size: inherit;
    height: 1em;
    overflow: visible;
    vertical-align: -0.125em;
}
```

juliusv commented:
(compare https://deploy-preview-8604--prometheus-react.netlify.app/graph and https://prometheus.demo.do.prometheus.io/graph and switch between the two tabs to see the difference)

prymitive commented:
This is caused by [reboot classes](https://github.com/twbs/bootstrap/blob/v4.6.0/scss/_reboot.scss#L257-L262) being included in both themes in https://github.com/ForEvolve/bootstrap-dark
Reboot is to reset all browsers defaults to same styles, but here it overrides some styles applied by FontAwesome, might need to customise ForEvolve/bootstrap-dark builds to remove it from themes and apply only once for root documents.

prymitive commented:
Added a workaround for reboot issues

juliusv commented:
:+1: Nice, thanks for filing that issue as well!

I'll merge this for now, and then we can individually file follow-up improvement PRs to the colors.

Thanks again for all the work! :)

prymitive commented:
Feel free to assign me tickets for any spotted issues.

juliusv commented:
@prymitive FYI https://twitter.com/juliusvolz/status/1382732267743809538
